File: examples/lua/hooks/replay.lua
-- ABOUTME: Lua example demonstrating hook replay functionality with parameter modification
-- ABOUTME: Shows how to use the Replay API for debugging and what-if analysis

-- Example 1: Create replay configurations
print("=== Replay Configuration Example ===")

-- Create different replay modes
local exactMode = Replay.modes.exact
local modifiedMode = Replay.modes.modified
local simulateMode = Replay.modes.simulate
local debugMode = Replay.modes.debug

print("Exact mode:", exactMode:name())
print("Modified mode:", modifiedMode:name())

-- Create a replay configuration
local config = Replay.create_config(modifiedMode)
print("Config mode:", config:get_mode():name())
print("Should compare results:", config:should_compare_results())
print("Timeout seconds:", config:get_timeout_seconds())

-- Example 2: Parameter modifications
print("\n=== Parameter Modifications Example ===")

-- Create parameter modifications
local mod1 = Replay.create_modification("context.data.input_value", 42.5, true)
local mod2 = Replay.create_modification("context.data.multiplier", 2.0, true)

print("Modification 1 path:", mod1:get_path())
print("Modification 1 value:", mod1:get_value())
print("Modification 1 enabled:", mod1:is_enabled())

-- Add modifications to config
config:add_modification("context.data.test_value", "modified", true)
config:add_modification("context.data.count", 100, true)

local mods = config:get_modifications()
print("Total modifications:", #mods)

-- Example 3: Replay schedules
print("\n=== Replay Schedules Example ===")

-- Create different schedule types
local onceSchedule = Replay.schedules.once(5.0)  -- 5 seconds delay
local intervalSchedule = Replay.schedules.interval(10.0, 60.0, 5)  -- 10s initial, 60s interval, max 5
local cronSchedule = Replay.schedules.cron("hourly")

print("Once schedule type:", onceSchedule:type_name())
print("Interval schedule type:", intervalSchedule:type_name())
print("Cron schedule type:", cronSchedule:type_name())

-- Example 4: Result comparison
print("\n=== Result Comparison Example ===")

-- Create a comparator
local comparator = Replay.create_comparator()

-- Create two JSON structures to compare
local original = {
    status = "success",
    value = 42,
    items = {1, 2, 3},
    metadata = {
        timestamp = 1234567890,
        user = "test"
    }
}

local replayed = {
    status = "success", 
    value = 43,  -- Different
    items = {1, 2, 3, 4},  -- Different length
    metadata = {
        timestamp = 1234567891,  -- Different
        user = "test"
    }
}

-- Compare the results
local comparison = comparator:compare_json(original, replayed)

print("Results identical:", comparison.identical)
print("Similarity score:", comparison.similarity_score)
print("Summary:", comparison.summary)

if comparison.differences then
    print("Differences found:")
    for i, diff in ipairs(comparison.differences) do
        print(string.format("  [%d] %s: %s", i, diff.path, diff.description))
    end
end

-- Example 5: Complex nested comparison
print("\n=== Complex Comparison Example ===")

local complex1 = {
    agents = {
        {name = "agent1", config = {model = "gpt-4", temperature = 0.7}},
        {name = "agent2", config = {model = "claude", temperature = 0.5}}
    },
    workflow = {
        steps = {"init", "process", "complete"},
        timeout = 300
    }
}

local complex2 = {
    agents = {
        {name = "agent1", config = {model = "gpt-4", temperature = 0.8}},  -- Different temp
        {name = "agent2", config = {model = "claude", temperature = 0.5}}
    },
    workflow = {
        steps = {"init", "process", "validate", "complete"},  -- Extra step
        timeout = 300
    }
}

local complexComparison = comparator:compare_json(complex1, complex2)
print("Complex comparison - Identical:", complexComparison.identical)
print("Complex comparison - Similarity:", complexComparison.similarity_score)

-- Example 6: Working with modifications programmatically
print("\n=== Programmatic Modifications Example ===")

-- Create a config for what-if analysis
local whatIfConfig = Replay.create_config(Replay.modes.modified)

-- Add multiple modifications based on conditions
local testValues = {10, 20, 50, 100}
for i, value in ipairs(testValues) do
    whatIfConfig:add_modification(
        string.format("context.data.test_%d", i),
        value * 2,
        true
    )
end

print("Created", #whatIfConfig:get_modifications(), "what-if modifications")

-- Disable some modifications
local allMods = whatIfConfig:get_modifications()
if #allMods > 2 then
    allMods[2]:set_enabled(false)
    print("Disabled modification at index 2")
end

print("\nReplay Lua example completed!")

File: llmspell-bridge/src/lua/globals/replay.rs
// ABOUTME: Lua bindings for hook replay functionality including scheduling and comparisons
// ABOUTME: Provides scripting access to advanced replay capabilities for debugging

use crate::lua::conversion::lua_value_to_json;
use llmspell_core::LLMSpellError;
use llmspell_hooks::replay::{
    HookResultComparator, ParameterModification, ReplayConfig, ReplayMode, ReplaySchedule,
};
use mlua::{
    Error as LuaError, FromLua, Lua, Result as LuaResult, Table, UserData, UserDataMethods, Value,
};
use serde_json;
use std::time::Duration;
use tracing::{instrument, trace};

/// Lua wrapper for `ReplayMode`
#[derive(Debug, Clone)]
pub struct LuaReplayMode(pub ReplayMode);

impl<'lua> FromLua<'lua> for LuaReplayMode {
    fn from_lua(value: Value<'lua>, _lua: &'lua Lua) -> LuaResult<Self> {
        match value {
            Value::UserData(ud) => Ok(ud.borrow::<Self>()?.clone()),
            _ => Err(LuaError::FromLuaConversionError {
                from: value.type_name(),
                to: "LuaReplayMode",
                message: Some("expected LuaReplayMode userdata".to_string()),
            }),
        }
    }
}

impl UserData for LuaReplayMode {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method("name", |_, this, ()| match this.0 {
            ReplayMode::Exact => Ok("exact"),
            ReplayMode::Modified => Ok("modified"),
            ReplayMode::Simulate => Ok("simulate"),
            ReplayMode::Debug => Ok("debug"),
        });
    }
}

/// Lua wrapper for `ParameterModification`
#[derive(Debug, Clone)]
pub struct LuaParameterModification(pub ParameterModification);

impl UserData for LuaParameterModification {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method("get_path", |_, this, ()| Ok(this.0.path.clone()));

        methods.add_method("get_value", |lua, this, ()| {
            let json_str = serde_json::to_string(&this.0.value).map_err(LuaError::external)?;
            lua.load(&json_str).eval::<Value>()
        });

        methods.add_method("is_enabled", |_, this, ()| Ok(this.0.enabled));

        methods.add_method_mut("set_enabled", |_, this, enabled: bool| {
            this.0.enabled = enabled;
            Ok(())
        });
    }
}

/// Lua wrapper for `ReplayConfig`
#[derive(Debug, Clone)]
pub struct LuaReplayConfig(pub ReplayConfig);

impl UserData for LuaReplayConfig {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method("get_mode", |_, this, ()| Ok(LuaReplayMode(this.0.mode)));

        methods.add_method("should_compare_results", |_, this, ()| {
            Ok(this.0.compare_results)
        });

        methods.add_method("get_timeout_seconds", |_, this, ()| {
            Ok(this.0.timeout.as_secs())
        });

        methods.add_method("should_stop_on_error", |_, this, ()| {
            Ok(this.0.stop_on_error)
        });

        methods.add_method("get_modifications", |_, this, ()| {
            Ok(this
                .0
                .modifications
                .clone()
                .into_iter()
                .map(LuaParameterModification)
                .collect::<Vec<_>>())
        });

        methods.add_method_mut(
            "add_modification",
            |_lua, this, (path, value, enabled): (String, Value, Option<bool>)| {
                let json_value = lua_value_to_json(value)?;
                this.0.modifications.push(ParameterModification {
                    path,
                    value: json_value,
                    enabled: enabled.unwrap_or(true),
                });
                Ok(())
            },
        );
    }
}

/// Lua wrapper for `ReplaySchedule`
#[derive(Debug, Clone)]
pub struct LuaReplaySchedule(pub ReplaySchedule);

impl UserData for LuaReplaySchedule {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method("type_name", |_, this, ()| match &this.0 {
            ReplaySchedule::Once { .. } => Ok("once"),
            ReplaySchedule::At { .. } => Ok("at"),
            ReplaySchedule::Interval { .. } => Ok("interval"),
            ReplaySchedule::Cron { .. } => Ok("cron"),
        });
    }
}


/// Inject replay global into Lua environment
///
/// # Errors
///
/// Returns an error if:
/// - Replay API creation fails
/// - Global injection fails
#[instrument(level = "trace", skip(lua), fields(global_name = "Replay"))]
pub fn inject_replay_global(lua: &Lua) -> Result<(), LLMSpellError> {
    trace!("Injecting Replay global API");
    let replay = create_replay_api(lua).map_err(|e| LLMSpellError::Internal {
        message: e.to_string(),
        source: None,
    })?;

    lua.globals()
        .set("Replay", replay)
        .map_err(|e| LLMSpellError::Internal {
            message: e.to_string(),
            source: None,
        })?;

    Ok(())
}

/// Create the replay API table
///
/// # Errors
///
/// Returns an error if:
/// - Table creation fails
/// - Method injection fails
pub fn create_replay_api(lua: &Lua) -> LuaResult<Table<'_>> {
    let replay = lua.create_table()?;

    // ReplayMode constructors
    let modes = lua.create_table()?;
    modes.set("exact", LuaReplayMode(ReplayMode::Exact))?;
    modes.set("modified", LuaReplayMode(ReplayMode::Modified))?;
    modes.set("simulate", LuaReplayMode(ReplayMode::Simulate))?;
    modes.set("debug", LuaReplayMode(ReplayMode::Debug))?;
    replay.set("modes", modes)?;

    // Create a default config
    replay.set(
        "create_config",
        lua.create_function(|_, mode: Option<LuaReplayMode>| {
            let config = ReplayConfig {
                mode: mode.map_or(ReplayMode::Exact, |m| m.0),
                ..Default::default()
            };
            Ok(LuaReplayConfig(config))
        })?,
    )?;

    // Create parameter modification
    replay.set(
        "create_modification",
        lua.create_function(
            |_lua, (path, value, enabled): (String, Value, Option<bool>)| {
                let json_value = lua_value_to_json(value)?;
                Ok(LuaParameterModification(ParameterModification {
                    path,
                    value: json_value,
                    enabled: enabled.unwrap_or(true),
                }))
            },
        )?,
    )?;

    // Create schedule types
    let schedules = lua.create_table()?;

    schedules.set(
        "once",
        lua.create_function(|_, delay_seconds: f64| {
            Ok(LuaReplaySchedule(ReplaySchedule::Once {
                delay: Duration::from_secs_f64(delay_seconds),
            }))
        })?,
    )?;

    schedules.set(
        "interval",
        lua.create_function(
            |_, (initial_delay, interval, max_executions): (f64, f64, Option<usize>)| {
                Ok(LuaReplaySchedule(ReplaySchedule::Interval {
                    initial_delay: Duration::from_secs_f64(initial_delay),
                    interval: Duration::from_secs_f64(interval),
                    max_executions,
                }))
            },
        )?,
    )?;

    schedules.set(
        "cron",
        lua.create_function(|_, expression: String| {
            Ok(LuaReplaySchedule(ReplaySchedule::Cron { expression }))
        })?,
    )?;

    replay.set("schedules", schedules)?;

    // Create comparator
    replay.set(
        "create_comparator",
        lua.create_function(|_, ()| Ok(LuaHookResultComparator))?,
    )?;

    Ok(replay)
}

/// Lua wrapper for `HookResultComparator`
#[derive(Debug)]
pub struct LuaHookResultComparator;

impl UserData for LuaHookResultComparator {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method(
            "compare_json",
            |lua, _this, (original, replayed): (Value, Value)| {
                let orig_json = lua_value_to_json(original)?;
                let repl_json = lua_value_to_json(replayed)?;

                // Convert to HookResult for comparison
                let orig_result = llmspell_hooks::HookResult::Modified(orig_json);
                let repl_result = llmspell_hooks::HookResult::Modified(repl_json);

                let comparator = HookResultComparator::new();
                let comparison = comparator.compare(&orig_result, &repl_result);

                // Convert comparison result to Lua table
                let result = lua.create_table()?;
                result.set("identical", comparison.identical)?;
                result.set("similarity_score", comparison.similarity_score)?;
                result.set("summary", comparison.summary)?;

                if !comparison.differences.is_empty() {
                    let diffs = lua.create_table()?;
                    for (i, diff) in comparison.differences.iter().enumerate() {
                        let diff_table = lua.create_table()?;
                        diff_table.set("path", diff.path.clone())?;
                        diff_table.set("description", diff.description.clone())?;
                        diffs.set(i + 1, diff_table)?;
                    }
                    result.set("differences", diffs)?;
                }

                Ok(result)
            },
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_lua_replay_api_creation() {
        let lua = Lua::new();
        let result = create_replay_api(&lua);
        assert!(result.is_ok());
    }
}
